json_rpc/clients/httpclient.nim

```diff
@@ -14,188 +16,45 @@ type
-    debug "Problems with networking", address = transp.remoteAddress(),
-          error = exc.msg
-    return ""
-
-  if not transp.validateResponse(header):
-    return ""
-
-  let length = header.contentLength()
-  if length > maxBodySize:
-    debug "Request body too large", length, maxBodySize
-    return ""
-
-  try:
-    if length > 0:
-      # `Content-Length` is present in response header.
-      buffer.setLen(length)
-      let blenfut = transp.readExactly(addr buffer[0], length)
-      let ores = await withTimeout(blenfut, HttpBodyTimeout)
-      if not ores:
-        # Timeout
-        debug "Timeout expired while receiving request body",
-              address = transp.remoteAddress()
-        return ""
-
-      blenfut.read() # exceptions
-    else:
-      # `Content-Length` is not present in response header, so we are reading
-      # everything until connection will be closed.
-      var blenfut = transp.read(maxBodySize)
-      let ores = await withTimeout(blenfut, HttpBodyTimeout)
-      if not ores:
-        # Timeout
-        debug "Timeout expired while receiving request body",
-              address = transp.remoteAddress()
-        return ""
-
-      buffer = blenfut.read()
-  except TransportIncompleteError:
-    # remote peer disconnected
-    debug "Remote peer disconnected", address = transp.remoteAddress()
-    return ""
-  except TransportOsError as exc:
-    debug "Problems with networking", address = transp.remoteAddress(),
-          error = exc.msg
-    return ""
-
-  return string.fromBytes(buffer)
 
 proc new(T: type RpcHttpClient, maxBodySize = MaxHttpRequestSize): T =
   T(
     maxBodySize: maxBodySize,
-    options: HttpClientOptions(httpMethod: MethodPost),
+    httpSession: HttpSessionRef.new(),
   )
 
 proc newRpcHttpClient*(maxBodySize = MaxHttpRequestSize): RpcHttpClient =
   RpcHttpClient.new(maxBodySize)
 
-proc httpMethod*(client: RpcHttpClient): HttpMethod =
-  client.options.httpMethod
-
-proc httpMethod*(client: RpcHttpClient, m: HttpMethod) =
-  client.options.httpMethod = m
-
 method call*(client: RpcHttpClient, name: string,
-             params: JsonNode): Future[Response] {.
-    async, gcsafe, raises: [Defect, CatchableError].} =
-  ## Remotely calls the specified RPC method.
-  let id = client.getNextId()
+             params: JsonNode): Future[Response]
+            {.async, gcsafe, raises: [Defect, CatchableError].} =
+  doAssert client.httpSession != nil
+  if client.httpAddress.isErr:
+    raise newException(RpcAddressUnresolvableError, client.httpAddress.error)
 
   let
-    transp = await connect(client.addresses[0])
+    id = client.getNextId()
     reqBody = $rpcCallNode(name, params, id)
-    res = await transp.sendRequest(reqBody, client.httpMethod)
+    req = HttpClientRequestRef.post(client.httpSession,
+                                    client.httpAddress.get,
+                                    body = reqBody.toOpenArrayByte(0, reqBody.len - 1))
+    res = await req.send()
```

Comment:
This could raise `HttpError`.

Reply:
https://github.com/bernardoaraujor/nim-json-rpc/commit/98d79570f83f743c756177e7f0b49cdc56c7a2c3